security/pkg/nodeagent/caclient/providers/keyfactor/client.go

```diff
@@ -0,0 +1,260 @@
+			}
+			customMetadatas[aliasName] = serviceName
+		case "PodName":
+			customMetadatas[aliasName] = metadata.PodName
+		case "PodNamespace":
+			customMetadatas[aliasName] = metadata.PodNamespace
+		case "PodIP":
+			customMetadatas[aliasName] = metadata.PodIP
+		case "TrustDomain":
+			customMetadatas[aliasName] = metadata.TrustDomain
+		}
+	}
+
+	c := &KeyFactorCAClient{
+		CaEndpoint:      endpoint,
+		EnableTLS:       enableTLS,
+		CustomMetadatas: customMetadatas,
+		Metadata:        metadata,
+		ClientOptions:   keyfactorConfig,
+	}
+
+	if !enableTLS {
+
+		transport := &http.Transport{
+			TLSClientConfig: &tls.Config{
+				InsecureSkipVerify: true,
+			},
+		}
+		c.Client = &http.Client{
+			Transport: transport,
+			Timeout:   30 * time.Second,
+		}
+	} else {
+
+		// Load the system default root certificates.
+		pool, err := x509.SystemCertPool()
+		if err != nil {
+			keyFactorCAClientLog.Errorf("could not get SystemCertPool: %v", err)
+			return nil, fmt.Errorf("could not get SystemCertPool: %v", err)
+		}
+
+		if pool == nil {
+			log.Info("System cert pool is nil, create a new cert pool")
+			pool = x509.NewCertPool()
+		}
+
+		if len(rootCert) > 0 {
+			ok := pool.AppendCertsFromPEM(rootCert)
+			if !ok {
+				return nil, fmt.Errorf("invalid root-cert.pem: %v", string(rootCert))
+			}
+		}
+
+		tlsConfig := &tls.Config{
+			RootCAs: pool,
+		}
+
+		transport := &http.Transport{TLSClientConfig: tlsConfig}
+
+		c.Client = &http.Client{
+			Transport: transport,
+			Timeout:   30 * time.Second,
+		}
+	}
+
+	return c, nil
+}
+
+// CSRSign calls KeyFactor CA to sign a CSR.
+func (cl *KeyFactorCAClient) CSRSign(ctx context.Context, reqID string, csrPEM []byte, subjectID string,
+	certValidTTLInSec int64) ([]string /*PEM-encoded certificate chain*/, error) {
+
+	bytesRepresentation, err := json.Marshal(keyfactorRequestPayload{
+		CSR:                  string(csrPEM),
+		CertificateAuthority: cl.ClientOptions.CaName,
+		IncludeChain:         true,
+		Template:             cl.ClientOptions.CaTemplate,
+		TimeStamp:            time.Now().Format(time.RFC3339),
+		Metadata:             cl.CustomMetadatas,
+	})
+
+	if err != nil {
+		keyFactorCAClientLog.Errorf("error encode json data: %v", err)
+		return nil, fmt.Errorf("error encode json data: %v", err)
+	}
+
+	u, err := url.Parse(cl.CaEndpoint)
+
+	if err != nil {
+		keyFactorCAClientLog.Errorf("invalid caAddress: %v (%v)", cl.CaEndpoint, err)
+		return nil, fmt.Errorf("invalid caAddress: %v (%v)", cl.CaEndpoint, err)
+	}
+
+	u.Path = path.Join(u.Path, cl.ClientOptions.EnrollPath)
+	enrollCSRPath := u.String()
+
+	keyFactorCAClientLog.Infof("start sign Keyfactor CSR request to: %v", enrollCSRPath)
+	requestCSR, err := http.NewRequest("POST", enrollCSRPath, bytes.NewBuffer(bytesRepresentation))
+
+	if err != nil {
+		return nil, fmt.Errorf("cannot create request with url: %v", enrollCSRPath)
+	}
+
+	requestCSR.Header.Set("authorization", cl.ClientOptions.AuthToken)
```

Comment:
So you set the header for authorization using the authtoken. From the code, you load the authtoken from a file. So for this external CA. Do you have a plan that instead of using the authtoken(which may expires or fails), use the already generated certificate as an authorization method to do CSR to your external CA to get or refresh a new certificate? 

I have this question because we try to implement a feature in the istiod CA server and may change the CSRSign Interface

Reply:
@williamaronli In the RFC @JDKeyfactor linked previously we proposed moving the specific Keyfactor auth/credentials outside of Istio and instead will pass through the JWT token from K8S similar to other integrations.